### PR TITLE
[libc++][modules] Rewrite the modulemap to have fewer top-level modules

### LINKED ISSUE #86193: [libc++] Per-header header modules
Commit 571178a21a8bc105bf86cf4bf92f842e07792e1a introduces per header file module with a main motivation to enable detection of cyclic includes. This means that now, instead of having a single binary artifact `std.pcm` or similar we get hundreds of pcm files for the same content. That's problematic for us due to several reasons:
  * Makes modules redistribution harder -- we carry build artifacts across nodes and it makes it difficult to enumerate all stl-related modules including private ones.
  * Efficiency -- we care about binary size and in-memory footprint. Going for a many pcm files has an overhead. Consider we have several textual files (such as cassert). The textual content becomes duplicated in every `pcm` file and in all dependent `pcm` files too.

If the motivation is only to be able to diagnose cyclic includes we can do that as part of the build process of libc++. For example, `echo "#include <libcxx/header>" | clang -xc++ -fsyntax-only -` will be a sufficient check ensuring header consistency. Alternatively, we could have a "debug" modulemap with that structure.

If there is more motivation to the one I understood, I'd like to suggest having a separate modulemap, called `std.modulemap` which keeps the headers into a single pcm file and clients like us can use it instead. Note that the `module std [system] {  header "__std_clang_module"  export * }` does not work for us because it triggers building of all gazillion pcms.

@ian-twilightcoder, @ldionne, depending on the direction we get in this issue, I can work on creating the `std.modulemap` and opening a pull request.

cc: @hahnjo, @dpiparo, @ChuanqiXu9.

### PR DESCRIPTION
**Pull request by ldionne**
This is a re-application of bc6bd3bc1e9 which was reverted in f11abac6524 because it broke the Clang pre-commit CI.

Original commit message:

This patch rewrites the modulemap to have fewer top-level modules. Previously, our modulemap had one top level module for each header in the library, including private headers. This had the well-known problem of making compilation times terrible, in addition to being somewhat against the design principles of Clang modules.

This patch provides almost an order of magnitude compilation time improvement when building modularized code (certainly subject to variations). For example, including <ccomplex> without a module cache went from 22.4 seconds to 1.6 seconds, a 14x improvement.

To achieve this, one might be tempted to simply put all the headers in a single top-level module. Unfortunately, this doesn't work because libc++ provides C compatibility headers (e.g. stdlib.h) which create cycles when the C Standard Library headers are modularized too. This is especially tricky since base systems are usually not modularized: as far as I know, only Xcode 16 beta contains a modularized SDK that makes this issue visible. To understand it, imagine we have the following setup:

   // in libc++'s include/c++/v1/module.modulemap
   module std {
      header stddef.h
      header stdlib.h
   }

   // in the C library's include/module.modulemap
   module clib {
      header stddef.h
      header stdlib.h
   }

Now, imagine that the C library's <stdlib.h> includes <stddef.h>, perhaps as an implementation detail. When building the `std` module, libc++'s <stdlib.h> header does `#include_next <stdlib.h>` to get the C library's <stdlib.h>, so libc++ depends on the `clib` module.

However, remember that the C library's <stdlib.h> header includes <stddef.h> as an implementation detail. Since the header search paths for libc++ are (and must be) before the search paths for the C library, the C library ends up including libc++'s <stddef.h>, which means it depends on the `std` module. That's a cycle.

To solve this issue, this patch creates one top-level module for each C compatibility header. The rest of the libc++ headers are located in a single top-level `std` module, with two main exceptions. First, the module containing configuration headers (e.g. <__config>) has its own top-level module too, because those headers are included by the C compatibility headers.

Second, we create a top-level std_core module that contains several dependency-free utilities used (directly or indirectly) from the __math subdirectory. This is needed because __math pulls in a bunch of stuff, and __math is used from the C compatibility header <math.h>.

As a direct benefit of this change, we don't need to generate an artificial __std_clang_module header anymore to provide a monolithic `std` module, since our modulemap does it naturally by construction.

A next step after this change would be to look into whether math.h really needs to include the contents of __math, and if so, whether libc++'s math.h truly needs to include the C library's math.h header. Removing either dependency would break this annoying cycle.

Thanks to Eric Fiselier for pointing out this approach during a recent meeting. This wasn't viable before some recent refactoring, but wrapping everything (except the C headers) in a large module is by far the simplest and the most effective way of doing this.

Fixes #86193

### CODE REVIEW

File: libcxx/include/module.modulemap [hunk 1]
```diff
@@ -1,2125 +1,2235 @@
...
+    module add_lvalue_reference                       { header "__type_traits/add_lvalue_reference.h" }
+    module add_pointer                                { header "__type_traits/add_pointer.h" }
+    module add_rvalue_reference                       { header "__type_traits/add_rvalue_reference.h" }
+    module add_volatile                               { header "__type_traits/add_volatile.h" }
+    module aligned_storage                            { header "__type_traits/aligned_storage.h" }
+    module aligned_union                              { header "__type_traits/aligned_union.h" }
+    module alignment_of                               { header "__type_traits/alignment_of.h" }
+    module can_extract_key                            { header "__type_traits/can_extract_key.h" }
+    module common_reference                           { header "__type_traits/common_reference.h" }
+    module common_type                                { header "__type_traits/common_type.h" }
```
Comment (frederick-vs-ja) [inline libcxx/include/module.modulemap:74]: As the CI failure is related to `common_type`, should we export this?

Comment (ldionne) [reply libcxx/include/module.modulemap:74]: Yeah, I wanted to reproduce the CI issue first. What I tried for now is this, which solves the issue: https://github.com/llvm/llvm-project/pull/110501/commits/ca58c03ec8244ac0f0f74e646b7f8fe6b6a0ce60.

However after seeing your comment just now I tried adding `export *` to the `common_type` module and that works too, so I think I'll go for that instead.

**Comment by ldionne** [comment]
Alright, I reproduced the issue that broke the Clang CI, I understand it and I validated that it was fixed in the revision right before this last update. I had to rebase to fix a conflict but there are no additional changes. I'll land this now.

**Comment by ldionne** [comment]
Please ping here and/or on Discord if anyone runs into issues with this change.

**Comment by frederick-vs-ja** [comment]
@ldionne The current organization seems problematic for implementing P1317R2, which brought several tuple-like mechanisms into `<type_traits>`. See #151480. Should we (largely) expand the `std_core` module, or create some non-core modules for `<type_traits>`?
(Also, it's quite confusing that `type_traits` is in core but `concepts` is not.)

**Pull request merged**